docs/RFCS/20211130_node_id_refactoring.md

```diff
@@ -0,0 +1,115 @@
+- Feature Name: Refactoring KV and SQL node iDs
+- Status: draft
+- Start Date: 2021-11-31
+- Authors: Radu Berinde
+- RFC PR: (PR # after acceptance of initial draft)
+- Cockroach Issue: (one or more # from the issue tracker)
+
+
+# Background
+
+Originally, CockroachDB nodes always ran both SQL and KV; this assumption is
+implicit in a lot of existing code but it is no longer true with multi-tenant
+support, where a tenant can have SQL nodes that are not KV nodes.
+
+Existing code used `roachpb.NodeID` everywhere to refer to a node, whether it is
+in the context of a KV RPC or a SQL (eg DistSQL) RPC. With multi-tenant support,
+a new type `base.SQLInstanceID` was added. Various pieces of the code were
+refactored to use this ID instead. However, there is a lot of remaining code
+that uses `NodeID` to refer to either a SQL instance or a KV node. Importantly,
+it is very easy for a new engineer to add subtly incorrect code.
+
+# Summary
+
+This is a proposal for implementing these IDs in a type-safe way, which
+disallows "unsafe" implicit conversions.
+
+# Proposal
+
+## Types
+
+We propose using three types, as defined below. Note that the chosen names are
+only for the purposes of this RFC (it is intentional that they are different
+than any of the names currently used in the code). We will establish the final
+names in the Naming section.
+
+```go
+// KVNodeID refers to a KV node.
+// Valid IDs start with 1.
+type KVNodeID int32
+
+func (id KVNodeID) String() string {
+  return fmt.Sprintf("kv%d", id)
+}
+
+// SQLNodeID refers to a SQL instance. This can be an instance inside a single
+// tenant or multi-tenant host cluster node or a standalone non-system tenant
+// instance. Note that this type already exists as `SQLInstanceID`.
+// Valid IDs start with 1.
+type SQLNodeID int32
+
+func (id KVNodeID) String() string {
+  return fmt.Sprintf("sql%d", id)
+}
+
+
+// GenericNodeID contains either a KVNodeID or a SQLNodeID. It is meant to be
+// used in code that is used for both cases and is generally agnostic as to what
+// the ID means (e.g. logging code).
+// The zero value is not a valid ID. The internal integer value is positive for
+// KVNodeIDs and negative for SQLNodeIDs.
+type GenericNodeID int32
```

Comment:
Have we considered the alternative of using an interface that both KVNodeID and SQLNodeID would implement.  This RFC is concerned with unsafe implicit conversions between KVNodeID and SQLNodeID but why do we also want people to have to explicitly convert into a generic ID when calling a function that takes a generic ID?